server/storage/backend/batch_tx_test.go

```diff
@@ -205,3 +207,110 @@ func TestBatchTxBatchLimitCommit(t *testing.T) {
 		return nil
 	})
 }
+
+func TestRangeAfterDeleteBucketMatch(t *testing.T) {
+	b, _ := betesting.NewTmpBackend(t, time.Hour, 10000)
+	defer betesting.Close(t, b)
+
+	tx := b.BatchTx()
+
+	tx.Lock()
+	tx.UnsafeCreateBucket(schema.Test)
+	tx.UnsafePut(schema.Test, []byte("foo"), []byte("bar"))
+	tx.Unlock()
+	tx.Commit()
+
+	checkRangeResponseMatch(t, b.BatchTx(), b.ReadTx(), []byte("foo"), nil, 0, [][]byte{[]byte("foo")}, [][]byte{[]byte("bar")})
+	checkForEachResponseMatch(t, b.BatchTx(), b.ReadTx(), schema.Test, [][]byte{[]byte("foo")}, [][]byte{[]byte("bar")})
+
+	tx.Lock()
+	tx.UnsafeDeleteBucket(schema.Test)
+	tx.Unlock()
+
+	checkForEachResponseMatch(t, b.BatchTx(), b.ReadTx(), schema.Test, nil, nil)
+
+	ksRead, _ := b.ReadTx().UnsafeRange(schema.Test, []byte("foo"), nil, 0)
+	if len(ksRead) > 0 {
+		t.Errorf("want no keys from ReadTx, got %v", ksRead)
+	}
+}
+
+func TestRangeAfterDeleteMatch(t *testing.T) {
+	b, _ := betesting.NewTmpBackend(t, time.Hour, 10000)
+	defer betesting.Close(t, b)
+
+	tx := b.BatchTx()
+
+	tx.Lock()
+	tx.UnsafeCreateBucket(schema.Test)
+	tx.UnsafePut(schema.Test, []byte("foo"), []byte("bar"))
+	tx.Unlock()
+	tx.Commit()
+
+	checkRangeResponseMatch(t, b.BatchTx(), b.ReadTx(), []byte("foo"), nil, 0, [][]byte{[]byte("foo")}, [][]byte{[]byte("bar")})
+	checkForEachResponseMatch(t, b.BatchTx(), b.ReadTx(), schema.Test, [][]byte{[]byte("foo")}, [][]byte{[]byte("bar")})
+
+	tx.Lock()
+	tx.UnsafeDelete(schema.Test, []byte("foo"))
+	tx.Unlock()
+	checkRangeResponseMatch(t, b.BatchTx(), b.ReadTx(), []byte("foo"), nil, 0, nil, nil)
+	checkForEachResponseMatch(t, b.BatchTx(), b.ReadTx(), schema.Test, nil, nil)
+}
+
+func checkRangeResponseMatch(t *testing.T, tx backend.BatchTx, rtx backend.ReadTx, key, endKey []byte, limit int64, expectedKeys, expectedValues [][]byte) {
```

Comment:
By adding validation of expected values you mixed assertion based testing with property based testing. Let's pick one of them.

Originally I implemented the test as property testing. The property I was testing was "ReadTx should always return the same response as BatchTX". This property should always hold, no matter what operations are executed before. I did it to allow make validation independent from test scenario and potentially randomize operations before validation.

Recent changes to the PR added `expectedKeys` and `expectedValues` testing, this means you hardeded the validation to specific cases. This is still correct, however the code can be much more simplified.

Reply:
In unit test, the test cases are hard coded. I think it is important to test the actual key and values in case ReadTx and BatchTX make the same mistake. 
Changed the code to just check the results in `checkForEach`